### ISSUE TITLE
noWindowExample issue in latest

### ISSUE THREAD

**Issue by dimitre** [openframeworks/openFrameworks#7781]
noWindow example fails compiling and running use make in macOS.
cc @artificiel 
```
/bin/sh: line 1: 27411 Segmentation fault: 11  ./noWindowExample
```

Crash Report
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   noWindowExample               	       0x1026f8eec std::__1::mersenne_twister_engine<unsigned int, 32ul, 624ul, 397ul, 31ul, 2567483615u, 11ul, 4294967295u, 7ul, 2636928640u, 15ul, 4022730752u, 18ul, 1812433253u>::operator()() + 0
1   noWindowExample               	       0x10271efc0 ofRandom(float) + 28
2   noWindowExample               	       0x1026adcc4 ofApp::draw() + 48
3   noWindowExample               	       0x1026ebf30 std::__1::__function::__func<std::__1::shared_ptr<of::priv::Function<ofEventArgs, std::__1::recursive_mutex>> ofEvent<ofEventArgs, std::__1::recursive_mutex>::make_function<ofBaseApp>(ofBaseApp*, void (ofBaseApp::*)(ofEventArgs&), int)::'lambda'(void const*, ofEventArgs&), std::__1::allocator<std::__1::shared_ptr<of::priv::Function<ofEventArgs, std::__1::recursive_mutex>> ofEvent<ofEventArgs, std::__1::recursive_mutex>::make_function<ofBaseApp>(ofBaseApp*, void (ofBaseApp::*)(ofEventArgs&), int)::'lambda'(void const*, ofEventArgs&)>, bool (void const*, ofEventArgs&)>::operator()(void const*&&, ofEventArgs&) + 40
4   noWindowExample               	       0x10276e968 ofEvent<ofEventArgs, std::__1::recursive_mutex>::notify(ofEventArgs&) + 156
5   noWindowExample               	       0x10276c77c ofCoreEvents::notifyDraw() + 40
6   noWindowExample               	       0x1026e9450 ofMainLoop::loopOnce() + 224
7   noWindowExample               	       0x1026e9344 ofMainLoop::loop() + 56
8   noWindowExample               	       0x1026ae434 main + 236
9   dyld                          	       0x197427f28 start + 2236
```
  
Different error in XCode:
```
[notice ] 1
Assertion failed: (the_instance), function instance, file ofSingleton.h, line 51.
```

**Comment by artificiel** [comment]
interesting: the no-window example code for main went from:
```c++
	ofAppNoWindow window;
	ofSetupOpenGL(&window, 1024, 768, OF_WINDOW);
	ofRunApp(new ofApp());
```
to new flavour:
```c++
	auto window = std::make_shared<ofAppNoWindow>();
	ofRunApp(window, std::make_shared<ofApp>());
	ofRunMainLoop();
```

which looks logical and cleaner.

but the initialization of the new random engine is called in `of::priv::initutils()` which is triggered within `ofInit()`, which makes it more like a symptom than the actual problem: `ofInit()` is called by `ofSetupOpenGL()` — this means the new flavour does not go through `ofInit()`. 

1. should `ofInit()` and/or `of::priv::initutils()` be called for `ofAppNoWindow`? (there seems to be windowing stuff in there, but also non-windowing things).
2. if `ofInit()` and/or `of::priv::initutils()` are not to be called for `ofAppNoWindow`, where should the random engine initialization be situated?
3. the OF guidelines say not to use `assert` — there is such a call in the original `ofSingleton` implementation I based ofSingleton on, and I left it because it helps tracking uninitialized instances, and as mentioned elsewhere, the favoured pattern of explicit singleton construction (as opposed to lazy initialization upon first access) removes many criticism for singleton implementations. here we see it provides good info, as the singleton is found to be uninitialized before being actually used and causing a plain memory access crash (if you want to see the crash you can try in Release which ignores `assert()`). anyway I just mention this as I don't have much experience with `assert` so I can't argue for/against , but I find this situation interesting.

TLDR: `ofRandomEngine()` is uninitialized because it's `::construct()` in `of::priv::initutils()` does not run.

**Comment by danoli3** [comment]
<img width="721" alt="Screenshot 2024-02-10 at 11 50 47 am" src="https://github.com/openframeworks/openFrameworks/assets/830748/eb385314-892a-467f-9027-3cf0c490aa59">

ofInit() in the ctor for this window type fixed it!
